### ISSUE TITLE
Feature suggestion: string replace transforms

### ISSUE THREAD

**Issue by halostatue** [fish-shell/fish-shell#7835]
I’m playing with a script and have run into an interesting encoding issue that I can’t handle with `string replace` and will have to pipe to an external program to do, but I think _could_ be handled in some cases with `string replace`. Here’s what I’m trying to do:

```sh
$ echo shurcoo!l | ruby -ne 'puts $_.gsub(/!(.)/) { $1.upcase }'
shurcooL
```

In Vim, I would do this with a regex like `s/!(.)/\u\1/`, so a form of this would be potentially useful for `string replace`:

```fish
echo shurcoo!l | string replace -r '!(.)' '\u$1'
```

**Comment by halostatue** [comment]
Here’s the special characters recognized by Vim (pulled from `:help sub-replace-special`):

* `\&` or `\0`: replaced with the whole matched pattern
* `\1`…`\9`: handled with `$1`…`$9` in Fish
* `\~`: replaced with the {string} of the previous substitute (I’m not sure what this would do)
* `\u`: next character made uppercase
* `\U`: following characters made uppercase, until `\E`
* `\l`: next character made lowercase
* `\L`: following characters made lowercase, until `\E`
* `\e` or `\E`: end of \u, \U, \l and \L (NOTE: not <Esc>!)
* `\r`: insert a carriage return
* `\n`: insert a newline
* `\b`: insert a backspace
* `\t`: insert a tab
* `\\`: insert a backslash
* `\={expr}`: insert the result of `expr`.

`\L\u` can be used to capitalize the first letter of a word.  This is not compatible with Vi and older versions of Vim, where the `\u` would cancel out the `\L`. Same for `\U\l`.

The last, (`\={expr}`) could be _really_ interesting: `string replace -r '!(.)' '\=(string upper $1)'`

**Comment by faho** [comment]
Almost all of these are already implemented:

>\& or \0: replaced with the whole matched pattern

`$0` is the whole matched pattern here.

>\u: next character made uppercase
\U: following characters made uppercase, until \E
\l: next character made lowercase
\L: following characters made lowercase, until \E
\e or \E: end of \u, \U, \l and \L (NOTE: not !)
\r: insert a carriage return
\n: insert a newline
\b: insert a backspace
\t: insert a tab
\\: insert a backslash

These all work as described - `string replace -r 'a(.*)' '\L$0\E' abc` prints "ABC".

>\~: replaced with the {string} of the previous substitute (I’m not sure what this would do)

This one doesn't really make sense.

>\={expr}: insert the result of expr.

This would appear to require vim's text editing DSL, which I don't think we want to implement.

So, in short:

Everything here that looks sensible to implement is already implemented.

In particular:

```fish
echo shurcoo!l | string replace -r '!(.)' '\u$1'
```

prints "shurcooL" on my system.

**Comment by halostatue** [comment]
Odd. That didn’t work for me. Can I suggest that these flags be documented as well?

```
❯ fish --version; echo shurcoo!l | string replace -r '!(.)' '\u$1'
fish, version 3.2.0
shurcool
```

**Comment by faho** [comment]
Fish relies on pcre2 to do this, I would suspect your OS's version is out of date.

**Comment by halostatue** [comment]
Hm. Frustrating, because I’m on pcre2 10.36 via Homebrew, but the Homebrew bottled version doesn’t appear to be support it. The Homebrew project is so hostile to issues that I’m not going to bother reporting this. ☹ (It’s not clear to me whether this is a problem with the configuration of Fish in Homebrew or of PCRE2 in Homebrew. Regardless, it’s frustrating.)

Thanks for the help.

**Comment by rouge8** [comment]
It might be a PCRE2 in Homebrew issue? It doesn't work for me building fish from source with brew installed PCRE2.

**Comment by halostatue** [comment]
> It might be a PCRE2 in Homebrew issue? It doesn't work for me building fish from source with brew installed PCRE2.

Yeah. I tried that. I don’t want to build PCRE2 from Homebrew to try this, either. It’s _deeply_ frustrating to deal with Homebrew oddness like this because they’re just _awful_ to report bugs to (if you don’t do it _exactly_ their way, the bug is auto-closed, and it take _way_ too long to do it _exactly_ their way, because it’s optimized for them and not for bug reporters; I get why because they get a lot of _dumb_ reports, but…).

**Comment by halostatue** [comment]
So `pcre2test` works on the Homebrew bottle:

```
printf "%s\n%s\n" '/!(.)/replace=\u$1,substitute_extended' shurcoo!l | pcre2test
PCRE2 version 10.36 2020-12-04
/!(.)/replace=\u$1,substitute_extended
shurcoo!l
 1: shurcooL
```

**Comment by halostatue** [comment]
Building from source on macOS Big Sur (non-Xcode; see #7838 for Xcode build issues) produces some warnings during the link step of the build:

```
Scanning dependencies of target fish_key_reader
[ 84%] Building CXX object CMakeFiles/fish_key_reader.dir/src/fish_key_reader.cpp.o
[ 85%] Building CXX object CMakeFiles/fish_key_reader.dir/src/print_help.cpp.o
[ 86%] Linking CXX executable fish_key_reader
ld: warning: dylib (/usr/local/lib/libpcre2-32.dylib) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: dylib (/usr/local/lib/libintl.dylib) was built for newer macOS version (10.15) than being linked (10.9)
[ 86%] Built target fish_key_reader
Scanning dependencies of target fish_indent
[ 87%] Building CXX object CMakeFiles/fish_indent.dir/src/fish_indent.cpp.o
[ 87%] Building CXX object CMakeFiles/fish_indent.dir/src/print_help.cpp.o
[ 88%] Linking CXX executable fish_indent
ld: warning: dylib (/usr/local/lib/libpcre2-32.dylib) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: dylib (/usr/local/lib/libintl.dylib) was built for newer macOS version (10.15) than being linked (10.9)
[ 88%] Built target fish_indent
Scanning dependencies of target fish
[ 89%] Building CXX object CMakeFiles/fish.dir/src/fish.cpp.o
[ 90%] Linking CXX executable fish
ld: warning: dylib (/usr/local/lib/libpcre2-32.dylib) was built for newer macOS version (11.0) than being linked (10.9)
ld: warning: dylib (/usr/local/lib/libintl.dylib) was built for newer macOS version (10.15) than being linked (10.9)
```

And the resulting version of Fish does not work, either:

```
$ ./fish -c 'string replace -r "!(.)" \'\\u$1\' "shurcoo!l" # => shurcooL'
shurcool
```

I can open these link issues as a separate issue, but I’m ultimately still trying to understand why this default feature—that works in `pcre2test` isn’t working.

**Comment by faho** [comment]
Ah, alright, I know what's happening here.

`string replace` had an error in the initial implemenation where it did an additional, superfluous, round of escaping on the replacement expression. That means any `\\` in there needs to be tripled, so you'll have to use

```fish

We have a feature flag `regex-easyesc` that you can turn on now, and we'll eventually turn it on by default and then force it on.
